### PR TITLE
upd redetect exception context

### PR DESCRIPTION
**Pull request by gluhar2006**

### CODE REVIEW

File: lunavl/sdk/detectors/facedetector.py [hunk 1]
```diff
@@ -272,16 +272,16 @@ def redetect(
...
         faces = []
         for image in images:
             assertImageForDetection(image.image)
             faces.extend(_createCoreFaces(image))
-        error, detectRes, _ = self._detector.redetect(
+        mainError, detectRes, errors = self._detector.redetect(
             faces, self._getDetectionType(detect5Landmarks, detect68Landmarks)
         )
-        assertError(error)
+        assertError(mainError, errors)
```
Comment (matemax) [inline lunavl/sdk/detectors/facedetector.py]: + test

Comment (gluhar2006) [reply lunavl/sdk/detectors/facedetector.py]: added

File: lunavl/sdk/detectors/humandetector.py [hunk 2]
```diff
@@ -213,13 +213,22 @@ def redetect(self, images: List[ImageForRedetection]) -> List[List[Union[HumanDe
...
+            isDetectionSuccess = True
             for bBox in image.bBoxes:
-                detection = self.redetectOne(image.image, bBox=bBox)
-                imageRes.append(detection)
-            res.append(imageRes)
+                try:
+                    detection = self.redetectOne(image.image, bBox=bBox)
+                    imageRes.append(detection)
+                except LunaSDKException as exc:
+                    errors.append(exc.error)
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/humandetector.py:243]: .

File: lunavl/sdk/detectors/humandetector.py [hunk 3]
```diff
@@ -213,13 +213,22 @@ def redetect(self, images: List[ImageForRedetection]) -> List[List[Union[HumanDe
...
+                    isDetectionSuccess = False
```
Comment (matemax) [inline lunavl/sdk/detectors/humandetector.py]: I think that one image one error
```python
for bBox in image.bBoxes:
    try:
        detection = self.redetectOne(image.image, bBox=bBox)
        imageRes.append(detection)
    except LunaSDKException as exc:
         errors.append(exc.error)
         break
    else:
         errors.append(Error.Succes)
else:
    res.append(imageRes)

File: lunavl/sdk/errors/exceptions.py [hunk 4]
```diff
@@ -50,15 +50,18 @@ def wrap(*func_args, **func_kwargs):
...
         error: optional
+        context: list with LunaSdkExceptions
 
     Raises:
         LunaSDKException: if optional contains error
 
     """
     if error.isError:
-        raise LunaSDKException(LunaVLError.fromSDKError(error))
+        if isinstance(error, FSDKErrorValueBool):
```
Comment (matemax) [inline lunavl/sdk/errors/exceptions.py]: what?

**Review by matemax — APPROVED** [review]

File: lunavl/sdk/detectors/humandetector.py (hunk 3 above)
Comment (gluhar2006) [reply lunavl/sdk/detectors/humandetector.py]: updated

File: lunavl/sdk/errors/exceptions.py (hunk 4 above)
Comment (gluhar2006) [reply lunavl/sdk/errors/exceptions.py]: removed

File: lunavl/sdk/detectors/humandetector.py [hunk 5]
```diff
@@ -139,10 +139,13 @@ def detectOne(
             forDetection = ImageForDetection(image=image, detectArea=image.rect)
         else:
             forDetection = ImageForDetection(image=image, detectArea=detectArea)
-        detections = self.detect([forDetection], limit=1, detectLandmarks=detectLandmarks)
-        if detections[0]:
-            return detections[0][0]
-        return None
+        imgs, detectAreas = getArgsForCoreDetectorForImages([forDetection])
+        error, detectRes = self._detector.detect(
+            [imgs[0]], [detectAreas[0]], 1, self._getDetectionType(detectLandmarks)
```
Comment (Zeskbest) [inline lunavl/sdk/detectors/humandetector.py:144]: i o not understand, what is `1` here

make it keyworded

File: lunavl/sdk/detectors/humandetector.py [hunk 6]
```diff
@@ -213,13 +228,23 @@ def redetect(self, images: List[ImageForRedetection]) -> List[List[Union[HumanDe
...
+                    detection = self.redetectOne(image.image, bBox=bBox)
+                    imageRes.append(detection)
+                    errors.append(LunaVLError.Ok.format(LunaVLError.Ok.description))
+                except LunaSDKException as exc:
+                    errors.append(exc.error)
+                    errorDuringProgress = True
+                    break
             res.append(imageRes)
+        if errorDuringProgress:
+            raise LunaSDKException(LunaVLError.BatchedInternalError, errors)
```
Comment (Zeskbest) [inline lunavl/sdk/detectors/humandetector.py:248]: there is no batching. Why `LunaVLError.BatchedInternalError` is raised?

File: lunavl/sdk/errors/exceptions.py [hunk 7]
```diff
@@ -50,15 +50,16 @@ def wrap(*func_args, **func_kwargs):
     return realWarp
 
 
-def assertError(error: FSDKErrorResult) -> None:
+def assertError(error: FSDKErrorResult, context: Optional[List[Any]] = None) -> None:
     """
     Assert core optional.
     Args:
         error: optional
+        context: list with LunaSdkExceptions
```
Comment (Zeskbest) [inline lunavl/sdk/errors/exceptions.py]: This is typing, not a comment

File: tests/base.py [hunk 8]
```diff
@@ -38,6 +38,18 @@ def assertLunaVlError(exceptionInfo: ExceptionInfo, expectedError: ErrorInfo):
...
+    def assertReceivedAndRawExpectedErrors(receivedError: ErrorInfo, expectedErrorWithoutDesc: ErrorInfo):
+        """
+        Assert expected and received errors as dicts
+        Args:
+            receivedError: received error
+            expectedErrorWithoutDesc: expected error without description (expected same as detail)
+        """
+        assert expectedErrorWithoutDesc.errorCode == receivedError.errorCode
+        assert expectedErrorWithoutDesc.description == receivedError.description
+        assert expectedErrorWithoutDesc.description == receivedError.detail
```
Comment (Zeskbest) [inline tests/base.py]: expectedErrorWithoutDesc.description  - is it empty? rename pls

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py [hunk 9]
```diff
@@ -62,8 +62,26 @@ def estimateDescriptorsBatch(
         tuple of batch and the aggregate descriptors (or None)
     Raises:
         LunaSDKException: if estimation failed
-
     """
+
+    def collectErrorByExtractingOneByOne():
+        """Extract descriptor without batching and collect errors"""
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/estimators_utils/extractor_utils.py]: add raises field

File: tests/test_estimate_descriptor.py [hunk 10]
```diff
@@ -377,6 +382,31 @@ def test_extract_descriptors_batch_positive(self):
...
+                                    extractor.estimateDescriptorsBatch(
+                                        [case.warps[0], badWarp], aggregate=aggregate, **kw
+                                    )
+                                assert len(exceptionInfo.value.context) == 2, "Expect two errors in exception context"
+                                self.assertReceivedAndRawExpectedErrors(exceptionInfo.value.context[0], LunaVLError.Ok)
+                                self.assertReceivedAndRawExpectedErrors(
+                                    exceptionInfo.value.context[1], LunaVLError.InvalidImageSize
+                                )
+
     @unittest.skip("dont do it FSDK-2186")
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]: unskip pls

File: tests/test_estimate_descriptor.py [hunk 11]
```diff
@@ -377,6 +382,31 @@ def test_extract_descriptors_batch_positive(self):
                                 else:
                                     assert descriptorAggregated is None
 
+    def test_extract_descriptors_batch_positive_and_negative(self):
+        """
+        Test estimate descriptor batch with one good warp and one bad (expected error).
+        """
+        for case in self.cases:
+            with self.subTest(type=case.type):
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]: one subtest pls

File: tests/test_facedetector.py [hunk 12]
```diff
@@ -7,9 +7,7 @@
 from lunavl.sdk.faceengine.setting_provider import DetectorType
 from lunavl.sdk.image_utils.geometry import Rect
 from lunavl.sdk.image_utils.image import VLImage, ColorFormat
-from tests.detect_test_class import FaceDetectTestClass
-from tests.resources import ONE_FACE, MANY_FACES, NO_FACES
-from tests.schemas import jsonValidator, REQUIRED_FACE_DETECTION, LANDMARKS5
+from tests.detect_test_class import FaceDetectTestClass, BAD_IMAGE
 from tests.detect_test_class import (
```
Comment (Zeskbest) [inline tests/test_facedetector.py:10]: why splitted?

File: tests/test_facedetector.py [hunk 13]
```diff
@@ -127,6 +127,17 @@ def test_batch_detect_using_different_type_detector(self):
                 detection = detector.detect(images=[VLIMAGE_ONE_FACE])[0]
                 self.assertFaceDetection(detection, VLIMAGE_ONE_FACE)
 
+    def test_batch_detect_with_success_and_error(self):
+        """
+        Test batch detection with success and error using FACE_DET_V3 (only need to check errors)
```
Comment (Zeskbest) [inline tests/test_facedetector.py]: why FACE_DET_V3 only?

File: lunavl/sdk/estimators/face_estimators/basic_attributes.py [hunk 14]
```diff
@@ -282,7 +282,17 @@ def estimateBasicAttributesBatch(
...
         error, baseAttributes, aggregateAttribute = self._coreEstimator.estimate(images, AttributeRequest(dtAttributes))
         if error.isError:
-            raise LunaSDKException(LunaVLError.fromSDKError(error))
+            errors = []
+            for image in images:
+                errorOne, baseAttributesOne = self._coreEstimator.estimate(image, AttributeRequest(dtAttributes))
+                if errorOne.isOk:
+                    errors.append(LunaVLError.Ok.format(LunaVLError.Ok.description))
+                else:
+                    errors.append(LunaVLError.fromSDKError(error))
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/face_estimators/basic_attributes.py]: errorOne

**Review by Zeskbest — CHANGES_REQUESTED** [review]

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py [hunk 15]
```diff
@@ -62,8 +62,26 @@ def estimateDescriptorsBatch(
...
+                descriptor = descriptorFactory.generateDescriptor()
+                coreDescriptor = descriptor.coreEstimation
+
+            errorOne, _ = coreEstimator.extractFromWarpedImage(warp.warpedImage.coreImage, coreDescriptor)
+            if errorOne.isError:
+                errors.append(LunaVLError.fromSDKError(errorOne))
+            else:
+                errors.append(LunaVLError.Ok.format(LunaVLError.Ok.description))
+
+        raise LunaSDKException(LunaVLError.BatchedInternalError, errors)
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/estimators_utils/extractor_utils.py]: just return errors here

File: lunavl/sdk/detectors/humandetector.py (hunk 5 above)
Comment (gluhar2006) [reply lunavl/sdk/detectors/humandetector.py:144]: unfortenatuly I cannot; comment added

File: lunavl/sdk/errors/exceptions.py (hunk 7 above)
Comment (gluhar2006) [reply lunavl/sdk/errors/exceptions.py]: upd

File: tests/base.py (hunk 8 above)
Comment (gluhar2006) [reply tests/base.py]: renamed

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py (hunk 9 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/estimators_utils/extractor_utils.py]: raising moved

File: tests/test_estimate_descriptor.py (hunk 10 above)
Comment (gluhar2006) [reply tests/test_estimate_descriptor.py]: ok

File: tests/test_estimate_descriptor.py (hunk 11 above)
Comment (gluhar2006) [reply tests/test_estimate_descriptor.py]: ok

File: tests/test_facedetector.py (hunk 12 above)
Comment (gluhar2006) [reply tests/test_facedetector.py:10]: pycharm

File: tests/test_facedetector.py (hunk 13 above)
Comment (gluhar2006) [reply tests/test_facedetector.py]: update description

File: lunavl/sdk/estimators/face_estimators/basic_attributes.py (hunk 14 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/basic_attributes.py]: fixed

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py (hunk 15 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/estimators_utils/extractor_utils.py]: noway

File: tests/test_facedetector.py [hunk 16]
```diff
@@ -127,6 +128,17 @@ def test_batch_detect_using_different_type_detector(self):
                 detection = detector.detect(images=[VLIMAGE_ONE_FACE])[0]
                 self.assertFaceDetection(detection, VLIMAGE_ONE_FACE)
 
+    def test_batch_detect_with_success_and_error(self):
+        """
+        Test batch detection with success and error using FACE_DET_V3 (there is not error with other detector)
```
Comment (Zeskbest) [inline tests/test_facedetector.py]:
```suggestion
        Test batch detection with success and error using FACE_DET_V3 (there is no error with other detector)
```

**Review by Zeskbest — APPROVED** [review]

**Pull request merged**